### PR TITLE
Optimize filter stage in TopNSorter

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Motivation
The `TopNSorter` consists of two parts: filter and merge-sort. The filter stage skips data rows that larger than the k-th row in current top-n, this procedure is pretty fast since it could finish in `O(N)` comparison. The merge-sort stage is performing a merge sort between baseline data and delta data, generate the new top-n.

This optimization targets at the first filter stage. 

## Implementation

To filter rows, A plain implementation is iterate all rows in chunk, and compare it with right row. While the problem is compare  overhead dominates, since it's a virtual function call.

So, we introduce the column-wise compare algorithm:
1. Compare the first left column with the first datum in right row, save result in `cmp_vector`
2. Compare the second left column only if the first left column is equal to first right value
3. repeat the second step until the last column

In this way, the compare procedure is conducted in column-wise style, which eliminated the `Column::compare_at` function call. And the compare result for each column is represented as a vector of int8_t, which takes only a little bit of memory footprint and could support fast navigate through SIMD instructions.

## Experiment Result
1. The experiment variables are data rows(16777216 to 67108864) , data columns(1 to 3), number of limit (8 to 262144)
2. The y-seris is sort speed, x-seris is different benchmark
3. The green line is before optimization, the blue line is after optimization
4. The result shows that, this optimization could achieve about 20% speedup when the data size is not so large

image

### CODE REVIEW

**Comment by mofeiatwork** [comment]
run all

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Review by Pslydhh — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by stdpain** [comment]
you could resolve them in next PR

File: be/src/exec/vectorized/sorting/compare_column.cpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    Status do_visit(const vectorized::NullableColumn& column) {
+        // Two step compare:
+        // 1. Compare null values, store at temporary result
+        // 2. Mask notnull values, and compare not-null values
+        const NullData& null_data = column.immutable_null_column_data();
+
+        // Set byte to 0 when it's null/null byte is 1
+        CompareVector null_vector(null_data.size());
+        for (size_t i = 0; i < null_data.size(); i++) {
+            null_vector[i] = (null_data[i] == 1) ? 0 : 1;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply be/src/exec/vectorized/sorting/compare_column.cpp:81]: good idea

**Pull request merged**